src/systems/optical_tactile_plugin/Visualization.cc

```diff
@@ -0,0 +1,239 @@
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_y(
+    _sensorPose.Rot().Y());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_z(
+    _sensorPose.Rot().Z());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_w(
+    _sensorPose.Rot().W());
+
+  this->node.Request("/marker", this->sensorMarkerMsg);
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginVisualization::VisualizeNormalForce(
+  uint64_t &_i, uint64_t &_j,
+  ignition::math::Vector3f &_position,
+  ignition::math::Vector3f &_normalForce,
+  ignition::math::Pose3f &_sensorWorldPose)
+{
+  // Visualization is downsampled according to plugin parameter
+  // <visualization_resolution> if <visualize_forces> is set to true
+
+  // Check if _normalForce has to be visualized
+  bool visualizeForce = (_i % this->visualizationResolution == 0 &&
+    _j % this->visualizationResolution == 0) ||
+    (_i % this->visualizationResolution == 0 && _j == 1) ||
+    (_i == 1 && _j % this->visualizationResolution == 0) ||
+    (_i == 1 && _j == 1);
+
+  if (!visualizeForce)
+    return;
+
+  if (_i == 1 && _j == 1)
+    this->markerID = 1;
+
+  // Add new markers with specific lifetime
+  this->positionMarkerMsg.set_id(this->markerID);
+  this->positionMarkerMsg.set_action(ignition::msgs::Marker::ADD_MODIFY);
+  this->positionMarkerMsg.mutable_lifetime()->set_sec(this->cameraUpdateRate);
+  this->positionMarkerMsg.mutable_lifetime()->set_nsec(0);
+
+  this->forceMarkerMsg.set_id(this->markerID++);
+  this->forceMarkerMsg.set_action(ignition::msgs::Marker::ADD_MODIFY);
+  this->forceMarkerMsg.mutable_lifetime()->set_sec(this->cameraUpdateRate);
+  this->forceMarkerMsg.mutable_lifetime()->set_nsec(0);
+
+  // The pose of the  markers must be published with reference
+  // to the simulation origin
+  ignition::math::Vector3f forceMarkerSensorPosition(
+    _position.X(), _position.Y(), _position.Z());
+
+  ignition::math::Quaternionf forceMarkerSensorQuaternion;
+  forceMarkerSensorQuaternion.From2Axes(
+    ignition::math::Vector3f(0, 0, 1), _normalForce);
+
+  // The position of the force marker is modified in order to place the
+  // end of the cylinder in the surface, not its middle point
+  ignition::math::Pose3f forceMarkerSensorPose(forceMarkerSensorPosition +
+    _normalForce * (this->forceLength/2), forceMarkerSensorQuaternion);
+
+  ignition::math::Pose3f forceMarkerWorldPose =
+    (forceMarkerSensorPose + this->depthCameraOffset) + _sensorWorldPose;
+  forceMarkerWorldPose.Correct();
+
+  ignition::math::Pose3f positionMarkerSensorPose(
+    forceMarkerSensorPosition, forceMarkerSensorQuaternion);
+  ignition::math::Pose3f positionMarkerWorldPose =
+    (positionMarkerSensorPose + this->depthCameraOffset) + _sensorWorldPose;
+  positionMarkerWorldPose.Correct();
+
+  // Set markers pose messages from previously computed poses
+  this->forceMarkerMsg.mutable_pose()->mutable_position()->set_x(
```

Comment:
The problem is that `msgs::Set()` are [implemented](https://ignitionrobotics.org/api/msgs/5.1/Utility_8hh.html) for `Vector3d`, and `Pose3d` and I am using `float`. `Color` can be changed though.

Reply:
Float can always be automatically promoted to double. It should let you do it.

Reply:
I meant I'm using `Vector3f`, `Quaternionf` and `Pose3f` (float) instead of `Vector3d`, `Quaterniond` and `Pose3d` (double), so I can't use `msgs::Set()` between those types

Reply:
The variables like `forceMarkerSensorPosition` in this method are all local scope, and you can use `Vector3d` etc for them. That will not take up much extra space since it's only a few variables, and they only exist for the duration of this method. With the messages being in double, you won't be saving much space anyway. That way, the following blocks will be a lot cleaner.

Reply:
e216bb2

I've `msgs::Set` the `Color` msgs, and for the remaining `Pose3d` msg needed I wrote [this](https://github.com/ignitionrobotics/ign-gazebo/blob/e216bb2132d0a77ca77dc31a50435487273fbe09/src/systems/optical_tactile_plugin/Visualization.cc#L79-L83). [Same](https://github.com/ignitionrobotics/ign-gazebo/blob/e216bb2132d0a77ca77dc31a50435487273fbe09/src/systems/optical_tactile_plugin/Visualization.cc#L181-L191) for the `Vector3d` msgs.